### PR TITLE
ENH: first detravisification step - an initial set of GitHub Actions for our Python and TwinCAT Projects

### PR DESCRIPTION
**Pull request by klauer**
# Note

This PR originally had a 'detravisify' tool included, but the scope shifted quickly toward writing reusable GH Actions for us to migrate to.

The migration tool and recent setup.py->pyproject.toml tool will live in a [separate repository](https://github.com/pcdshub/pcds-python-migration-tools)

This still represents the first step of moving away from Travis, so this PR and the branch name will remain.

# GitHub actions workflows
* Testing on pcdsutils/pcdsdevices separately
* Workflows for:
- [x] Conda build / test (combined into one job due to how the test environment is reused from `conda mambabuild`)
- [x] Pip install / test
- [x] Pip and docs-versions-menu documentation building and deployment
- [x] Pre-commit checks
- [x] PyPI uploading
- [x] Anaconda uploading
- [x] TwinCAT things

## Related PRs and Projects

* Python project cookiecutter PR to use this workflow (among many other updates):
    * https://github.com/pcdshub/cookiecutter-pcds-python/pull/37
* Python CI testing repository: https://github.com/pcdshub/pcds-ci-test-repo-python
    * [Generated docs](https://pcdshub.github.io/pcds-ci-test-repo-python/v0.0.2/)
* TwinCAT CI testing repository: https://github.com/pcdshub/pcds-ci-test-repo-twincat
    * [Generated docs](https://pcdshub.github.io/pcds-ci-test-repo-twincat/v0.0.1/)
    * PR which adds pragma lint: https://github.com/pcdshub/pcds-ci-test-repo-twincat/pull/2/files
* pcdsutils migration to GHA: https://github.com/pcdshub/pcdsutils/pull/69
    * Output: https://github.com/klauer/pcdsutils/actions/runs/3878848661
* Migration tool that takes our standard travis.yml and converts it to a workflow: https://github.com/pcdshub/pcds-python-migration-tools/pull/2

## Example project workflow files

(These are also now in the repository root as "example*gha.yml")

### Python

As recommended in the cookiecutter (`klauer` -> `pcdshub` and branch to be changed, of course)

```yaml

name: PCDS Standard Testing

on:
  push:
  pull_request:
  release:
    types:
      - created

jobs:
  standard:
    uses: klauer/pcds-ci-helpers/.github/workflows/python-standard.yml@enh_gha
    with:
      # The workflow needs to know the package name.  This can be determined
      # automatically if the repository name is the same as the import name.
      package-name: "{{ cookiecutter.import_name }}"
      # Extras that will be installed for both conda/pip:
      testing-extras: "line_profiler<4"
      # Extras to be installed only for conda-based testing:
      conda-testing-extras: "pyqt=5"
      # Extras to be installed only for pip-based testing:
      pip-testing-extras: "pyqt5"
      # Use setuptools-scm in the conda-build workflow
      use-setuptools-scm: true
```

### TwinCAT

```yaml
name: PCDS Standard Testing

on:
  push:
  pull_request:
  release:
    types:
      - created

jobs:
  standard:
    uses: klauer/pcds-ci-helpers/.github/workflows/twincat-standard.yml@enh_gha
    with:
      # The workflow can be configured to look at a specific project directory, if
      # desirable.  This defaults to the root of the repository (".").
      project-root: "."
      # Pattern (grep-compatible) of filenames to exclude in style checks.
      style-exclude: ""
```

## Notes

WIP notes

(TIL: Interestingly GitHub markdown allows triple-backticks within a quadruple-backtick section)

<details>

````
References:
1. https://docs.github.com/en/actions/using-workflows/workflow-syntax-for-github-actions
2. For local, greppable copy: https://github.com/github/docs/tree/main/content/actions

Running actions locally:

> https://github.com/nektos/act/

Limitations:

1. Ubuntu 22.04 incompatible with a lot of actions somehow
2. `act --graph` is neat:
  ```
  $ act --graph
   ╭───────╮
   │ build │
   ╰───────╯
      ⬇
   ╭──────╮
   │ test │
   ╰──────╯

  ```
3. Artifact support requires a server to be running to receive the artifacts.
``act`` auto-starts a server when you use `--artifact-server-path` (a local
path to store artifacts)

```
$ mkdir ./build_artifacts
$ act --artifact-server-path ./build_artifacts
```

4. With artifacts, GITHUB_RUN_ID matters. Run act with:

```
act --env GITHUB_RUN_ID=$(date '+%s')
```

5. As soon as development is done, migrate to reusable workflows. Then ``act``
   stops working, sadly.

```bash
$ act
Error: reusable workflows are currently not supported (see https://github.com/nektos/act/issues/826 for updates)
```

Learned:

1. Multiline strings

```
abc: >-
 cmd
  arg1
  arg2
```

Should evaluate to "cmd arg1 arg2" with special yaml syntax- https://yaml-multiline.info/
It does not, somehow. Don't use it.

2. Bash arrays in `env:` sections do not work. They become regular strings with parentheses.
  Don't bother trying to use them.  Splitting strings into bash arrays not worth it due to
  how complicated it is.

3. Circular dependencies

4. conda test - see notes in workflow

5. Setting per-step defaults is useful and saves typing

6. `act` does not work well with ubuntu 22.04 to install Python
  (actions/setup-python), but 20.04 is just fine.  Sticking with 20.04 for now.

7. caching conda packages between CI jobs/workflows:
  a. Notes indicate that this must be set in condarc:
    `use_only_tar_bz2: true`
  b. Failure with `KeyError: Dist(channel='conda-forge', dist_name='ca-certificates-2022.12.7-ha878542_0', name='ca-certificates', fmt='.conda', version='2022.12.7', build_string='ha878542_0', build_number=0, base_url=None, platform=None)`
     during mambabuild indicates cache problem, perhaps?
  c. Cache is quick, 300mb in 7 seconds or so. How long do package downloads take?
  d. Sometimes cache download step fails miserably and loops:
    `Received 313893288 of 318087592 (98.7%), 4.6 MBs/sec`
    - mamba + boa env built in 30 seconds
    - pcdsdevices env built in 1 min 30 seconds
  e. Overall conclusion: caching complexity and problems not worth it...

8. `export CONDA_PY=3.9` is the same as 'conda build --python 3.9'.  
    Recipes must have `python: {{PY_VER}}*,>=3.9` (e.g) in them for this to matter.
    However, building on 3.9 and installing on 3.11 results in:
    ``- package pcdsdevices-7.1.0+105.gd3d9bb094-py_1 requires python 3.9.*,>=3.9, but none of the providers can be installed``
    ... Turns out we need to set this as an extra-dep instead:
    ``--extra-deps "python==${{ matrix.python_version }}"``

9. matrix is per-job and not per-workflow
   build on 3.9 -> parallel matrix test 3.9, 3.10

10. `conda mambabuild --build-only` does not appear to generate pcdsdevices*.tar.bz2 in conda-bld
    Found this closed issue after discovering/resolving this on my own:
      https://github.com/conda/conda-build/issues/632

11. `conda mambabuild --no-test --no-anaconda-upload` generates the package as expected

12. The conda build can get even more dependencies from the command line via
    "--extra-deps" but I'd like to keep these in a separate job on CI, I think

13. Workflow reuse - env vars
  > "Any environment variables set in an env context defined at the workflow level
  > in the caller workflow are not propagated to the called workflow. For more
  > information about the env context, see 'Context and expression syntax for
  > GitHub Actions.'"

14. Workflow reuse - technically possible to layer things with "composite" actions:
```
  repo
    - conda-standard (has matrix)
        - conda-build 3.9 (composite action)
        - conda-build 3.10 (composite action)
        - conda-build 3.11 (composite action)
```
but output in GH actions GUI shows not so helpfully: 
  "running klauer/pcds-ci-helpers/github/actions/conda-build@enh_gha"
without individual steps

Doing it like this results in similarly bad output:

``` 
    repo (has matrix)
      - conda-build 3.9 (composite action)
      - conda-build 3.10 (composite action)
      - conda-build 3.11 (composite action)
``` 

15. version pins and bash escaping are very painful, but maybe I'm missing an easier solution

16. YAML anchors unsupported, meaning we have to repeat ourselves for things like
    the Python version matrix

    https://github.com/actions/runner/issues/1182

17. Workflow yaml files must be in the top-level ``.github/workflows`` directory:
```
invalid value workflow reference: workflows must be defined at the top level of the .github/workflows/ directory
``

18. setuptools-scm + conda is a pain! 

  a. setuptools-scm + pip install/build wheel/etc works well, as expected
  b. dynamic version checking also appears to work fine
  c. conda-build (+mambabuild) require that the version be set prior to
     the build process starting, though they offer the ability to load information
     from `setup.py` (which we don't want to use) or from an arbitrary file.
     There is no pre-build hook to insert this call.

  One option I've found to keep compatibility with versioneer workflows is
  using 'load_file_regex' in the per-project conda recipe.  This is flimsy
  and still doesn't quite work as it requires setuptools-scm to be called
  once prior to the recipe being built.  That necessitated an additional step
  in our CI pipeline - installing setuptools-scm and initializing _version.py

19. New documentation-building pipeline will squash (amend) commits between
  tags to avoid massive gh-pages branch. I think this is a pretty good approach
  since we never manually fix these branches.

````

</details>

#### TwinCAT Style Annotations

image

#### TwinCAT Pragma Lint Annotations

image

(Due to pytmc showing line numbers in terms of source code and not XML-tagged code, the line numbers are sadly a few off. This could be addressed in the future in that repository.)

#### flake8 problem matcher from the pre-commit job

image

### CODE REVIEW

**Comment by klauer** [comment]
This is ready for a first review.

If it passes the overall smell test and the features are found to be acceptable, I think migrating one repository and following up with a review of how well this worked could be a good next step.

**Review by tangkong — APPROVED** [review]
I think the output on pcdsutils looks great, I'm a sucker for the little job-flow-graph-thing.  I had a good chuckle at the twincat syle job, I didn't realize that's how we did/do it.

I think things look good here, though I'm interested to see how the jobs render failures in more detail.  

A tangential thought I had: I wonder if how the intermittent CI errors will manifest themselves on GHA, or if anything will be different at all.  In that vein I'd be down to transition lightpath over and see how GHA works on that CI as well.  It's a repo that probably won't see too much development in the near future?

**Comment by tangkong** [comment]
Also god bless your WIP notes section <3

File: .github/workflows/python-conda-test.yml [hunk 1]
```diff
@@ -0,0 +1,265 @@
+name: Conda Build
+
+on:
+  workflow_call:
+    inputs:
+      package-name:
+        default: ""
+        description: "The package name to be built"
+        required: false
```
[the remark this answers is not in the document]
Comment (klauer) [reply .github/workflows/python-conda-test.yml:9]: Ah, I'm glad you caught this one. That's a mistake.

It wasn't required originally as I had a "detect package method from organization name" step. I later took it out as auto-detection could be prone to error (i.e., if the package name and the organization name are different - for example with `-` being `_` in Python) and that it's easy enough for the user to specify.

It seems like a good input to require going forward and I wouldn't want to have to backtrack to add it to our jobs down the line. It's actually only a placeholder in the conda/docs job and only really used in the pip job at the moment:
https://github.com/klauer/pcds-ci-helpers/blob/00dc75fff1a7cc5d58a88368cb6f932627483c73/.github/workflows/python-pip-test.yml#L102-L114

File: .github/workflows/python-conda-test.yml [hunk 2]
```diff
@@ -0,0 +1,265 @@
...
+          --log-file="$HOME/logs/debug_log.txt" \
+          --log-format='%(asctime)s.%(msecs)03d %(module)-15s %(levelname)-8s %(threadName)-10s %(message)s' \
+          --log-file-date-format='%H:%M:%S' \
+          --log-level=DEBUG \
+          2>&1 | tee "$HOME/logs/pytest_log.txt"
+
+    - name: pcds-dev deployment
+      if: inputs.deploy-on-success && github.event.pull_request.merged == true && github.ref == 'refs/heads/master'
+      env:
+        ANACONDA_API_TOKEN: ${{ secrets.CONDA_UPLOAD_TOKEN_DEV }}
```
[the remark this answers is not in the document]
Comment (klauer) [reply .github/workflows/python-conda-test.yml:221]: Correct. We can set up per-repository secrets and per-organization secrets as well. 

I expect the conda ones should be organization-level.
For the PyPI ones, I've been generating per-package tokens to be a bit safer, but it's a bit of a pain.

File: .github/workflows/python-conda-test.yml [hunk 3]
```diff
@@ -0,0 +1,265 @@
...
+          echo '```'
+        ) | tee -a "$GITHUB_STEP_SUMMARY"
+
+        echo "## Debug log"
+        cat "$HOME/logs/debug_log.txt" || echo "Debug logfile not found?"
+
+    - name: Upload log file artifacts
+      uses: actions/upload-artifact@v3
+      with:
+        name: Python ${{ inputs.python-version }} - conda - testing log
```
[the remark this answers is not in the document]
Comment (klauer) [reply .github/workflows/python-conda-test.yml:263]: This isn't entirely redundant. It includes:
* Debug logs (`debug_log.txt`) from the pytest phase 
* Pytest output (`pytest_log.txt`)

**Comment by klauer** [comment]
> I think things look good here, though I'm interested to see how the jobs render failures in more detail.

As far as how failures render, I think it's something we can tweak to our liking as we go.
I just added flake8 annotations which should help us with contributors to pcdsdevices that have yet to figure out pre-commit: 
image

Admittedly these can be noisy - but they shouldn't be popping up to start with if you follow the development practices we recommend.

File: .github/workflows/twincat-style.yml [hunk 4]
```diff
@@ -0,0 +1,124 @@
...
+    steps:
+    - uses: actions/checkout@v3
+      with:
+        fetch-depth: 0
+
+    - name: Find all changed files
+      if: ${{ github.event_name == 'pull_request' }}
+      run: |
+        # Fancy method of saying:
+        # * What files changed from this PR (the current checked out HEAD)
```
Comment (ZLLentz) [inline .github/workflows/twincat-style.yml:60]: This is great

**Review by ZLLentz — APPROVED** [review]
I'd like to start trying this out on a few repos, it looks pretty great to me. The example configs end up being much simpler than the travis ones too (I think?).

We can even run both travis and gha on a few repos as a trial if we were cautious.

**Comment by ZLLentz** [comment]
If we merge this, we can point the gh workflows to `pcdshub` instead of `klauer`, right?

**Comment by klauer** [comment]
Yes, that's correct @ZLLentz - and no `@enh_gha` or other branch specifier will be required.

GitHub docs generally recommend versioning shared CI configurations (with tags like `v1`), but I expect continuing to just use master for us will suffice for a long while.

**Pull request merged**